### PR TITLE
feat(crosswalk_traffic_light_estimator): add new package for crosswalk traffic light signals estimation

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Description

Now, autoware supports and uses only vehicle traffic light signals, but I think if perception module detected or estimated not only vehicle traffic light signals but also crosswalk traffic light signals, behavior planning module could make better decision. (e.g. crosswalk module)

So, in this PR, I added `traffic_light_estimater` node in order to get crosswalk traffic light singals.

This new node subscribes

- classified vehicle traffic light signals(`autoware_auto_perception_msgs::msg::TrafficSignalArray`)
- map(`autoware_auto_mapping_msgs::msg::HADMapBin`)
- route(`autoware_auto_planning_msgs::msg::PathWithLaneId`)

and publishes

- estimated crosswalk and vehicle traffic light signals(`autoware_auto_perception_msgs::msg::TrafficSignalArray`)

#### Situation1
If traffic between pedestrians and vehicles is controlled by traffic signals, and signals of crosswalks intersecting **STRAIGHT** lanelets with `TrafficLight::GREEN` traffic state, the crosswalk traffic light signals prevent pedestrians from crossing, so in this situation, this node output crosswalk signals as `TrafficLight::RED`.

<div align="center">
  
</div>
<div align="center">
  
</div>

#### Situation2
If traffic between pedestrians and vehicles is controlled by traffic signals, and signals of crosswalks intersecting different turn direction lanelets (e.g. straight and left, left and right...) with `TrafficLight::GREEN` traffic state, the crosswalk traffic light signals prevent pedestrians from crossing, so in this situation, this node output crosswalk signals as `TrafficLight::RED`.

<div align="center">
  
</div>

## Related links

- https://github.com/autowarefoundation/autoware.universe/issues/937

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by satoshi-ota** [comment]
- [ ] fix typo estimater -> estimator
- [ ] write README

**Comment by mitsudome-r** [comment]
Are you still working on this PR? (Just checking because there were no updates for a month)

**Comment by satoshi-ota** [comment]
@mitsudome-r 
> Are you still working on this PR? (Just checking because there were no updates for a month)

Sorry for the late reply :bow: I had been testing the logic at odaiba, and confirmed that the node works as expected :+1: 
Now, I'm prepairing this PR for review.

File: perception/traffic_light_estimator/src/node.cpp [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+}  // namespace
+
+TrafficLightEstimatorNode::TrafficLightEstimatorNode(const rclcpp::NodeOptions & options)
+: Node("traffic_light_estimator", options)
+{
+  using std::placeholders::_1;
+
+  use_last_detect_color_ = this->declare_parameter("use_last_detect_color", true);
```
Comment (satoshi-ota) [inline perception/traffic_light_estimator/src/node.cpp:83]: @yukkysaito FYI: If we set `use_last_detect_color_` **true**, the node estimates pedestrian's traffic signal as RED not only when vehicle's traffic signal is detected as GREEN but also when detection results change GREEN to UNKNOWN.

File: perception/traffic_light_estimator/src/node.cpp [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        RCLCPP_ERROR(get_logger(), "%s", ex.what());
+        return;
+      }
+    }
+  }
+
+  conflicting_crosswalks_.clear();
+
+  for (const auto & route_lanelet : route_lanelets) {
+    constexpr int PEDESTRIAN_GRAPH_ID = 1;
```
Comment (yukkysaito) [inline perception/traffic_light_estimator/src/node.cpp:142]: [Q] what is the number? where is defined about the number?

File: perception/traffic_light_estimator/src/node.cpp (hunk 1 above)
Comment (yukkysaito) [reply perception/traffic_light_estimator/src/node.cpp:83]: when detection results change RED to UNKNOWN, does it estimate as UNKNOWN? only when GREEN to UNKNOW, it is detected as GREEN, right?

Comment (satoshi-ota) [reply perception/traffic_light_estimator/src/node.cpp:83]: YES. That's right !

Comment (yukkysaito) [reply perception/traffic_light_estimator/src/node.cpp:83]: I would like to make the name easier to understand. But I have no idea...

Comment (satoshi-ota) [reply perception/traffic_light_estimator/src/node.cpp:83]: OK :+1:  Let me think about it ...

File: perception/traffic_light_estimator/src/node.cpp (hunk 2 above)
Comment (satoshi-ota) [reply perception/traffic_light_estimator/src/node.cpp:142]: I reused the code from following package, but I don't know abot where does the definition exisit :pleading_face: 

https://github.com/autowarefoundation/autoware.universe/blob/ce6ccec0872481c136dc8df8fe4af08992ffee48/planning/behavior_velocity_planner/src/scene_module/crosswalk/manager.cpp#L61-L62

File: perception/traffic_light_estimator/src/node.cpp (hunk 1 above)
Comment (satoshi-ota) [reply perception/traffic_light_estimator/src/node.cpp:83]: @yukkysaito I thought about better param name, but I also have no idea... :pleading_face: 
If we describe the parameter's description clearly in the README , I think that we don't have to fix the parameter's name and remain as it is now, what do you think?

I write README in https://github.com/autowarefoundation/autoware.universe/pull/944/commits/138e9c4a8641fb6955bae77191256ce7fb3a5ba2

File: perception/traffic_light_estimator/src/node.cpp (hunk 2 above)
Comment (satoshi-ota) [reply perception/traffic_light_estimator/src/node.cpp:142]: I asked the experts of lanelet library about that, the number is index of Participants::Pedestrian in `overall_grapths` and there is no definition in the library.

https://github.com/autowarefoundation/autoware.universe/blob/ce6ccec0872481c136dc8df8fe4af08992ffee48/planning/route_handler/src/route_handler.cpp#L181-L189

**Comment by satoshi-ota** [comment]
TEST at kashiwanoha 

WHITE CROSSWALK: related pedestrian's traffic signal is **RED** (crosswalk module don't look)
RED CROSSWALK: related pedestrian's traffic signal is **UNKNOWN** (crosswalk module look and stop when objects cross the road)

- basic intersection

https://user-images.githubusercontent.com/44889564/176823330-95d4ca41-4521-4367-9180-e7998cdb7ad1.mp4

https://user-images.githubusercontent.com/44889564/176822787-31a558f3-4339-4618-9bee-95eb999314f4.mp4

- T intersection

https://user-images.githubusercontent.com/44889564/176822777-97be83d9-569a-4ee1-9294-f3842497333f.mp4

https://user-images.githubusercontent.com/44889564/176822794-d06cee42-0aac-4e00-8199-fa9f3160f788.mp4

File: perception/traffic_light_estimator/src/node.cpp (hunk 1 above)
Comment (yukkysaito) [reply perception/traffic_light_estimator/src/node.cpp:83]: @satoshi-ota Sounds good :+1:

File: perception/traffic_light_estimator/src/node.cpp (hunk 2 above)
Comment (yukkysaito) [reply perception/traffic_light_estimator/src/node.cpp:142]: I see. Thank you

**Comment by yukkysaito** [comment]
I wanna change the package name. how about `crosswalk_traffic_light_estimater`?

**Comment by satoshi-ota** [comment]
> @satoshi-ota I wanna change the package name. how about `crosswalk_traffic_light_estimater`?

@yukkysaito OK, I agree with you :+1: I'll fix it this PR and autoware_launch.

**Comment by satoshi-ota** [comment]
TEST at odaiba

WHITE CROSSWALK: related pedestrian's traffic signal is RED (crosswalk module don't look)
RED CROSSWALK: related pedestrian's traffic signal is UNKNOWN (crosswalk module look and stop when objects cross the road)

https://user-images.githubusercontent.com/44889564/176837098-ca27da87-3a80-4b70-8162-23e4374f9730.mp4


https://user-images.githubusercontent.com/44889564/176837104-763bc610-501c-4ae8-aca3-a209ac43ea5c.mp4


https://user-images.githubusercontent.com/44889564/176837113-29e51fef-bc10-4804-bd68-2dc990f76398.mp4


https://user-images.githubusercontent.com/44889564/176837122-3c74a038-5094-46a3-be90-180e0998b755.mp4

**Comment by satoshi-ota** [comment]
@yukkysaito I fixed package name in https://github.com/autowarefoundation/autoware.universe/pull/944/commits/9daaafbd7ff56ef20d0b44e3f1562e09a0fee4e2

**Review by yukkysaito — APPROVED** [review]
LGTM

**Pull request merged**